Terminal.Gui/Drivers/CursesDriver.cs

```diff
@@ -62,8 +65,49 @@ public override void AddRune (Rune rune)
 		public override void AddStr (ustring str)
 		{
 			// TODO; optimize this to determine if the str fits in the clip region, and if so, use Curses.addstr directly
-			foreach (var rune in str)
+			foreach (var rune in str) {
+				if (IsOutBounds (rune))
+					return;
 				AddRune (rune);
+			}
+		}
+
+		bool IsOutBounds (Rune rune)
+		{
```

Comment:
Thank you for this heroic effort, but I suspect this is the wrong place to put this, the result is too complicated.

There are two ways of properly solving this, one is to track the "last known position" in View (the Move method already calls Driver.Move) so we would keep "current-col" and "current-row", and then introduce a `View.AddRune` method there, which only needs to check in the View whether it is out of bounds or not, and update its own counter there.

And then we move the code from using `Driver.AddRune` to `view.AddRune`.

We should then add a big fat [Obsolete] to `Driver.AddRune` to discourage its use (and the same goes for `AddStr`).

An alternative design to not keep track of a per-view col/row position, is to have AddRune in the view take a column/row, and advance that on each view that does drawing manually.  But I suspect this is too cumbersome.

A third design option is to do what UI toolkits do and is to create a "drawing-context" upon entering draw, so draw code would be like this (more or less, this is from the top off my head):

```
void Draw (Rect rect)
{
    var context = GetDrawingContext ();
    context.AddRune ("h");
    context.AddString ("ello");
}

And the x/y coordinates would be on the context.    Personally, I think that keeping this on the View is an acceptable tradeoff.

Reply:
I think you can put that in the View so that it can control that flow. These are ideas that come with time :)